core/src/main/scala/org/broadinstitute/dsde/rawls/monitor/migration/WorkspaceMigrationMonitor.scala

```diff
@@ -55,236 +57,392 @@ object WorkspaceMigrationMonitor {
       .map(_ > 0)
 
 
-  final def schedule(workspace: Workspace): WriteAction[Unit] = workspaceMigrations
-    .map(_.workspaceId)
-    .insert(workspace.workspaceIdAsUUID)
-    .ignore
+  final def schedule(workspace: Workspace): ReadWriteAction[Unit] =
+    workspaceMigrations
+      .map(_.workspaceId)
+      .insert(workspace.workspaceIdAsUUID)
+      .ignore
+
+
+  final case class MigrationDeps(dataSource: SlickDataSource,
+                                 billingProject: RawlsBillingProject,
+                                 workspaceService: WorkspaceService,
+                                 storageService: GoogleStorageService[IO],
+                                 storageTransferService: GoogleStorageTransferService[IO])
+
 
+  type MigrateAction[T] = ReaderT[OptionT[IO, *], MigrationDeps, T]
 
-  final def timestampNow: IO[Timestamp] = IO.delay(Timestamp.valueOf(LocalDateTime.now))
+  object MigrateAction {
+    // lookup a value in the environment
+    final def asks[T](f: MigrationDeps => T): MigrateAction[T] =
+      ReaderT.ask[OptionT[IO, *], MigrationDeps].map(f)
+
+    // lift an IO action into the context of a MigrateAction
+    final def liftIO[A](ioa: IO[A]): MigrateAction[A] =
+      ReaderT.liftF(OptionT.liftF(ioa))
+
+    // empty action
+    final def unit: MigrateAction[Unit] =
+      ReaderT.pure()
+  }
+
+
+  def getMigrations(workspaceUuid: UUID): MigrateAction[Seq[WorkspaceMigration]] =
+    inTransaction { _ =>
+      workspaceMigrations
+        .filter(_.workspaceId === workspaceUuid)
+        .sortBy(_.id)
+        .result
+    }
+
+  // Read workspace migrations in various states, attempt to advance their state forward by one
+  // step and write the outcome of each step to the database.
+  final def migrate: MigrateAction[Unit] =
+    startMigration >>
+      claimAndConfigureGoogleProject >>
+      createTempBucket >>
+      issueWorkspaceBucketTransferJob >>
+      deleteWorkspaceBucket >>
+      createFinalWorkspaceBucket >>
+      issueTmpBucketTransferJob >>
+      deleteTemporaryBucket >>
+      MigrateAction.unit
+
+
+  final def startMigration: MigrateAction[Unit] =
+    for {
+      migration <- findMigration(_.started.isEmpty)
```

Comment:
If I understand your question correctly then yes - this will return a migration if it has been scheduled and it's not currently being processed anywhere else in the pipeline.
`started` is somewhat artificial but I wanted to include it so we could monitor when the actor started working on any particular migration. My main motivation is that it allows us to consider alternative scheduling strategies. For example, we may want to limit the number of concurrent migrations we run initially, just until we know things are more or less OK before we ramp up migration. We might want to prioritise work on certain workspaces. We might want to do something else entirely. The `started` entry is a simple flag that can be used to encode this and throttle the number of migrations that enter the pipeline at once.

Reply:
Right, but couldn't we just use whatever is the first step to notice that? (I guess nothing could stop us from adding steps before that one, but I just worry about us adding more complexity that we don't need)

Reply:
I prefer to have a step that has no real functionality to represent this rather than bundling that in with steps with migration logic.